src/cxx11.hpp

```diff
@@ -0,0 +1,865 @@
+class thread
+  {
+  private:
+    ::_MCF_thread* _M_thr = nullptr;
+
+    thread(const thread&) = delete;
+    thread& operator=(const thread&) = delete;
+
+  public:
+    constexpr thread() noexcept = default;
+
+    template<typename _Callable, typename... _Args,
+        typename ::std::enable_if<
+            ! ::std::is_same<typename ::std::decay<_Callable>::type, thread>::value
+          >::type* = nullptr>
+    explicit
+    thread(_Callable&& __callable, _Args&&... __args)
+      {
+        enum { __evn_0, __evn_constructed, __evn_cancelled };
+        using _My_invoker = _Invoke_decay_copy<_Callable, _Args...>;
+
+        struct _My_data
+          {
+            ::_MCF_event _M_event[1];
+            _My_invoker _M_invoker[1];
+
+            static
+            void
+            __thunk(::_MCF_thread* __thr)
+              {
+                _My_data* __my = (_My_data*) ::_MCF_thread_get_data(__thr);
+
+                // Check whether `_M_invoker` has been initialized. If its
+                // constructor failed, this thread shall exit immediately.
+                int __err = ::_MCF_event_await_change(__my->_M_event, __evn_0, nullptr);
+                if(__err == __evn_cancelled)
+                  return;
+
+                // Execute the user-defined procedure.
+                __MCF_ASSERT(__err == __evn_constructed);
+                __my->_M_invoker->__do_it();
+                __my->_M_invoker->~_My_invoker();
+              };
+          };
+
+        // Create the thread. User-defined data are initialized to zeroes.
+        this->_M_thr = ::_MCF_thread_new_aligned(_My_data::__thunk, alignof(_My_data),
+                                                 nullptr, sizeof(_My_data));
+        if(!this->_M_thr)
+          __MCF_THROW_SYSTEM_ERROR(resource_unavailable_try_again,
+                                   "_MCF_thread_new_aligned");
+
+        _My_data* const __my = (_My_data*) ::_MCF_thread_get_data(this->_M_thr);
+
+        __MCF_TRY {
+          // Construct `_M_invoker`.
+          ::new((void*) __my->_M_invoker) _My_invoker(__callable, __args...);
+        }
+        __MCF_CATCH(...) {
+          // Notify the thread about the failure.
+          ::_MCF_event_set(__my->_M_event, __evn_cancelled);
+          ::_MCF_thread_drop_ref(this->_M_thr);
+          __MCF_THROW();
+        }
+
+        // Notify the thread about the success.
+        ::_MCF_event_set(__my->_M_event, __evn_constructed);
+      }
+
+    thread(thread&& __other) noexcept
+      {
+        this->_M_thr = __other._M_thr;
+        __other._M_thr = nullptr;
+      }
+
+    thread&
+    operator=(thread&& __other) noexcept
+      {
+        if(this->_M_thr)
+          ::std::terminate();
+
+        this->_M_thr = __other._M_thr;
+        __other._M_thr = nullptr;
+
+        return *this;
+      }
+
+    ~thread() noexcept
+      {
+        if(this->_M_thr)
+          ::std::terminate();
+      }
+
+    using id = uint32_t;
```

Comment:
This is not conforming, because the standard requires it to be a class type.

Reply:
well, what about `std::hash`?